### ISSUE TITLE
Layout class takes up most of compile time.

### ISSUE THREAD

**Issue by nonhermitian** [Qiskit/qiskit-terra#1731]
### Informations

- **Qiskit Terra version**: master
- **Python version**: 3.7
- **Operating system**: OSX

### What is the current behavior?
In the `StochasticSwap` mapper the `Layout` class takes up most of the compile time:
```
605432191 function calls (605207568 primitive calls) in 210.245 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
 61117244   51.807    0.000   58.954    0.000 register.py:114(__hash__)
 18510976   32.015    0.000  156.546    0.000 _layout.py:59(__setitem__)
309541422   25.141    0.000   25.141    0.000 {built-in method builtins.isinstance}
  6530943   24.118    0.000   37.742    0.000 _layout.py:93(<listcomp>)
 37021952   18.371    0.000   27.602    0.000 _layout.py:78(_checktype)
 18907960    8.972    0.000   75.860    0.000 _layout.py:54(__getitem__)
  9695556    7.803    0.000   59.968    0.000 _layout.py:74(__delitem__)
61171486/61170829    7.155    0.000    7.158    0.000 {built-in method builtins.hash}
  6530943    4.839    0.000   47.727    0.000 _layout.py:92(__len__)
  6561568    4.528    0.000    4.528    0.000 {built-in method builtins.max}
 18907960    3.843    0.000   15.914    0.000 {function Layout.__getitem__ at 0xa244b8f28}
   440767    3.833    0.000  148.674    0.000 _layout.py:29(from_dict)
25439672/25439666    2.986    0.000   50.712    0.000 {built-in method builtins.len}
       87    1.859    0.021  202.938    2.333 <ipython-input-4-ef448761bc1a>:101(_layer_permutation)
   440108    1.782    0.000   13.735    0.000 <ipython-input-4-ef448761bc1a>:221(<listcomp>)
    74202    0.687    0.000    2.096    0.000 _dagcircuit.py:174(_add_wire)
  6541268    0.631    0.000    0.631    0.000 {method 'keys' of 'dict' objects}
   440108    0.516    0.000   21.816    0.000 _layout.py:170(swap)
   440771    0.404    0.000  149.187    0.000 _layout.py:22(__init__)
```

Compare to old compile chain, with old `swap_mapper` (here with a register fix I did):
```
        67077888 function calls (66140520 primitive calls) in 36.861 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
 16692324   13.613    0.000   15.475    0.000 register.py:114(__hash__)
   328458    3.407    0.000   15.457    0.000 _mapping.py:256(<listcomp>)
  9241913    3.324    0.000    3.324    0.000 register.py:97(__eq__)
       57    2.612    0.046   30.168    0.529 _mapping.py:131(layer_permutation)
16745878/16745229    1.870    0.000    1.874    0.000 {built-in method builtins.hash}
   760716    0.705    0.000    1.856    0.000 register.py:28(__init__)
  1280111    0.625    0.000    0.854    0.000 graph.py:646(nodes)
    58574    0.575    0.000    1.694    0.000 _dagcircuit.py:174(_add_wire)
   380098    0.508    0.000    3.719    0.000 _mapping.py:245(<listcomp>)
    55177    0.401    0.000    0.748    0.000 facts.py:503(deduce_all_facts)
   760413    0.377    0.000    1.358    0.000 _coupling.py:51(size)
   184000    0.357    0.000    0.357    0.000 {method 'normal' of 'mtrand.RandomState' objects}
   760716    0.356    0.000    0.356    0.000 {method 'match' of 're.Pattern' objects}
```

As you can see, the `Layout` makes the new swapper ~6x slower than the old one.

### What is the expected behavior?

The compiler should not be made slower than before.

### Suggested solutions
The mapper should use simpler data structures, e.g. NumPy arrays or the like, and then reconstitute them to the desired format when returning the new DAG.  Once done, we can then look to push the whole routine down to lower code.

**Comment by kdk** [comment]
@1ucian0 I just pushed a branch that's a start on this.

**Comment by ajavadia** [comment]
closed via #1736
